### DISCUSSION TITLE
[SOLVED] ValueError: [E002] Can't find factory for 'transformer' for language Italian (it)

### DISCUSSION THREAD

**Discussion by Panzerotto93** [explosion/spaCy#10975] [Help: Installation] [answered]
## How to reproduce the behaviour
Create docker image with:
- spacy==3.2.4
- spacy[transformers]==3.2.4
- transformers==4.17.0
- tokenizers==0.12.1
In requirements. I downloaded the model files and put them in a folder inside the docker image.
Calling spacy.load(path/to/model/it_nerIta_trf), will raise the error in the title.

## Your Environment

- **spaCy version:** 3.2.4
- **Platform:** Windows-10-10.0.19044-SP0
- **Python version:** 3.9.7
- **Pipelines:** it_nerIta_trf (0.0.0)

How do I solve this? I have no problems running the code locally on Windows, but the error is raised in a Linux Docker container. Thanks!

**Comment by polm** [comment, marked as answer]
Have you installed `spacy-transformers`? Not finding the `transformer` factory doesn't mean the model files can't be found, it means spaCy doesn't have any knowledge of Transformer functionality. If you install a model through `pip` then dependencies like `spacy-transformers` will be automatically installed, but if you call `spacy.load` that won't happen, so you need to insure you already have dependencies.

**Reply by Panzerotto93** [reply]
Yes, I have spacy, spacy[transformers] and transformers in my requirements. Also, if I try to import spacy_transformers it says it doesn't find tqdm, even if tqdm is in the requirements.

**Reply by polm** [reply]
What do you mean you have them "in [your] requirements"? The requirements.txt file is a list of things you should install, it is not a list of things that are currently installed. You can check what's installed with `pip freeze`. Can you share the output of that here?

The way you use `requirements.txt` is `pip install -r requirements.txt`, so if you're missing things you should use that.

**Reply by Panzerotto93** [reply]
I mean that when I create the docker image, I have a requirements.txt file that contains all the necessary libraries (spacy, spacy[transformers], etc.). From the log it looks like all the installations are OK, but then when I try to import the model I get the error above. Or, if I try to import spacy-transformera, I get an error about missing tqdm.
I'll post the pip freeze result asap.

**Reply by Panzerotto93** [reply]
@polm unfortunately I can't paste the result of pip freeze, as it does not work in the docker image somehow. I now moved to another error: whenever I import spacy_transformers in the docker image, I get:
`importlib.metadata.PackageNotFoundError: The 'tqdm>=4.27' distribution was not found and is required by this application`
These are my requirements when I create the docker file:
```
numpy==1.20.3
pandas==1.3.2
beautifulsoup4==4.10.0
openpyxl==3.0.7
rank-bm25==0.2.2
scikit-learn==1.0.2
stop-words==2018.7.23
langdetect==1.0.9
Unidecode==1.3.3
nltk==3.7
tqdm==4.62.1
spacy==3.2.4
tokenizers==0.12.1
transformers==4.17.0
spacy[transformers]==3.2.4
```
If I import tqdm I get no errors.
I have no problem running the code on my laptop, only getting these errors in the docker image.

**Reply by polm** [reply]
Huh, OK. 

When you try to run `pip freeze` in the Docker image, what happens? What error do you get?

How did you generate this `requirements.txt` file?

Can you try removing the `spacy==3.2.4` and `transformers==4.17.0` lines, leaving the `spacy[transformers]` line? `spacy[transformers]` should cover both of those.

If you can share your Dockerfile, or some parts of it, that may help.

**Reply by Panzerotto93** [reply]
> Huh, OK.
> 
> When you try to run `pip freeze` in the Docker image, what happens? What error do you get?
> 
> How did you generate this `requirements.txt` file?
> 
> Can you try removing the `spacy==3.2.4` and `transformers==4.17.0` lines, leaving the `spacy[transformers]` line? `spacy[transformers]` should cover both of those.
> 
> If you can share your Dockerfile, or some parts of it, that may help.

I tried running pip freeze with subprocess, I think that was the problem, but I have no access to the docker shell for policy reasons.

I generated the requirements by hand, maybe I'll try with pipreqs. 

Thank you, I'll try removing these packages. But I cant' share the Dockerfile, again for policy reasons.

I think the real problem here is that I have little control after creating the Docker image and every try takes approx 20 mins.

Thank you again for your help, I'll let you know!

EDIT: these should be the installed packages using spacy[transformers] only in requirements.txt:
MarkupSafe-2.1.1 Unidecode-1.3.3 beautifulsoup4-4.10.0 blis-0.7.7 catalogue-2.0.7 certifi-2022.6.15 charset-normalizer-2.0.12 click-8.0.4 cymem-2.0.6 et-xmlfile-1.1.0 filelock-3.7.1 huggingface-hub-0.8.0 jinja2-3.1.2 joblib-1.1.0 langcodes-3.3.0 langdetect-1.0.9 murmurhash-1.0.7 nltk-3.7 numpy-1.20.3 openpyxl-3.0.7 packaging-21.3 pandas-1.3.2 pathy-0.6.1 preshed-3.0.6 pydantic-1.8.2 pyparsing-3.0.9 python-dateutil-2.8.2 pytz-2022.1 rank-bm25-0.2.2 regex-2022.6.2 requests-2.28.0 scikit-learn-1.0.2 scipy-1.8.1 six-1.16.0 smart-open-5.2.1 soupsieve-2.3.2.post1 spacy-3.2.4 spacy-alignments-0.8.5 spacy-legacy-3.0.9 spacy-loggers-1.0.2 spacy-transformers-1.1.6 srsly-2.4.3 stop-words-2018.7.23 thinc-8.0.17 threadpoolctl-3.1.0 tokenizers-0.12.1 torch-1.11.0 tqdm-4.62.1 transformers-4.19.4 typer-0.4.1 urllib3-1.26.9 wasabi-0.9.1

EDIT2: removed spacy and transformers from requirements. Same error raised.

**Reply by polm** [reply]
If you have `spacy-transformers` installed and can verify that I'm not sure what's going on. To be clear the error you're getting is the E002 error, not the tqdm error? It would help if you pasted the full stack trace. 

Kind of a long shot, but can you run a small piece of code like this?

```
import spacy
import spacy_transformers # normally not necessary, but try this line

nlp = spacy.load("your_model")
```

**Reply by Panzerotto93** [reply]
> If you have `spacy-transformers` installed and can verify that I'm not sure what's going on. To be clear the error you're getting is the E002 error, not the tqdm error? It would help if you pasted the full stack trace.
> 
> Kind of a long shot, but can you run a small piece of code like this?
> 
> ```
> import spacy
> import spacy_transformers # normally not necessary, but try this line
> 
> nlp = spacy.load("your_model")
> ```

If I try:
```
import spacy
nlp = spacy.load("your_model")
```
I get the E002 error (I'll paste the full full stack trace asap), so I tried adding in between:
`import spacy_transformers`
and I get:
```
import spacy_transformers

File "/usr/local/lib/python3.9/site-packages/spacy_transformers/__init__.py", line 1, in <module>

from . import architectures

File "/usr/local/lib/python3.9/site-packages/spacy_transformers/architectures.py", line 6, in <module>

from .layers import TransformerModel, TransformerListener

File "/usr/local/lib/python3.9/site-packages/spacy_transformers/layers/__init__.py", line 1, in <module>

from .listener import TransformerListener

File "/usr/local/lib/python3.9/site-packages/spacy_transformers/layers/listener.py", line 4, in <module>

from ..data_classes import TransformerData

File "/usr/local/lib/python3.9/site-packages/spacy_transformers/data_classes.py", line 5, in <module>

from transformers.tokenization_utils import BatchEncoding

File "/usr/local/lib/python3.9/site-packages/transformers/__init__.py", line 30, in <module>

from . import dependency_versions_check

File "/usr/local/lib/python3.9/site-packages/transformers/dependency_versions_check.py", line 41, in <module>

require_version_core(deps[pkg])

File "/usr/local/lib/python3.9/site-packages/transformers/utils/versions.py", line 120, in require_version_core

return require_version(requirement, hint)

File "/usr/local/lib/python3.9/site-packages/transformers/utils/versions.py", line 107, in require_version

raise importlib_metadata.PackageNotFoundError(

importlib.metadata.PackageNotFoundError: The 'tqdm>=4.27' distribution was not found and is required by this application.

Try: pip install transformers -U or pip install -e '.[dev]' if you're working with git main

```

**Reply by polm** [reply]
Well that's weird. It definitely looks like you don't have tqdm installed, though you say you can import it without an error.

What is the output of this code?

```
import tqdm
print(tqdm.__version__)
```

**Reply by Panzerotto93** [reply]
> Well that's weird. It definitely looks like you don't have tqdm installed, though you say you can import it without an error.
> 
> What is the output of this code?
> 
> ```
> import tqdm
> print(tqdm.__version__)
> ```

```
import tqdm
print(' tqdm version: ' + str(tqdm.__version__))

tqdm version: 4.62.1
```

If I try spacy.load("it_nerIta_trf")

```
ValueError: [E002] Can't find factory for 'transformer' for language Italian (it). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).


Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, doc_cleaner, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentencizer, textcat, spancat, textcat_multilabel, it.lemmatizer
```

I downloaded the model from: https://huggingface.co/bullmount/it_nerIta_trf

**Reply by Panzerotto93** [reply]
Guess the problem is with importlib.metadata. Running 

```
import importlib.metadata as importlib_metadata
importlib_metadata.version('tqdm')
```
raises ModuleNotFoundError
tried with:
```
import importlib
importlib.sys.path.append('/usr/local/lib/python3.8/site-packages/')
import importlib.metadata as importlib_metadata
importlib_metadata.version('tqdm')
```
Same thing happens

**Reply by polm** [reply]
We have had weird problems with `importlib` before, but nothing like that, and I'm honestly not sure what could cause it. 

Kind of a long shot but maybe you could try changing your Python version to 3.9? Since importlib is a core library that might change something subtle in it. 

Besides that... I'm not sure I can offer much more help unless you can make an example we can reproduce somehow to investigate. It sounds like somthing is wrong with your virtualenv but hard to say what.

**Reply by Panzerotto93** [reply]
Problem was with the default configuration of docker image made by IT, in which all the dist-info folders were deleted. Thus, every time importlib searched for tqdm dist-info folder, it couldn't find it and raised the error. Just had to comment that line in the dockerfile to solve the problem.
Sorry for bothering you and thanks a lor for the help!

**Reply by polm** [reply]
Glad you figured it out, and thanks for letting us know what caused it!

**Reply by polm** [reply]
Also as a note, in this forum we don't edit the titles of threads to add `[SOLVED]` or similar, we just use the "mark answer" feature to mark something as "done". Sometimes it's a little weird (like when the first post isn't a solution), but that's OK, as is not marking the thread solved in any way - we also have an internal tracker.
